File: Java/Problem_012.java
public class Main {

    private static long countDivisors(long x) {
        long total = 1;

        int exp = 0;
        while (x % 2 == 0) {
            x /= 2;
            exp++;
        }
        if (exp > 0) total *= (exp + 1);

        for (long p = 3; p * p <= x; p += 2) {
            exp = 0;
            while (x % p == 0) {
                x /= p;
                exp++;
            }
            if (exp > 0) total *= (exp + 1);
        }

        if (x > 1) total *= 2;

        return total;
    }

    public static void main(String[] args) {
        final long threshold = 500;
        long n = 1;

        while (true) {
            long a, b;
            if (n % 2 == 0) {
                a = n / 2;
                b = n + 1;
            } else {
                a = n;
                b = (n + 1) / 2;
            }

            if (countDivisors(a) * countDivisors(b) > threshold) {
                System.out.println(n * (n + 1) / 2); // 76576500
                return;
            }
            n++;
        }
    }
}

File: Python/Problem_012.py
def count_divisors(x: int) -> int:
    total = 1

    exp = 0
    while x % 2 == 0:
        x //= 2
        exp += 1
    if exp:
        total *= (exp + 1)

    p = 3
    while p * p <= x:
        exp = 0
        while x % p == 0:
            x //= p
            exp += 1
        if exp:
            total *= (exp + 1)
        p += 2

    if x > 1:
        total *= 2

    return total


def solve(threshold: int = 500) -> int:
    n = 1
    while True:
        if n % 2 == 0:
            a, b = n // 2, n + 1
        else:
            a, b = n, (n + 1) // 2

        if count_divisors(a) * count_divisors(b) > threshold:
            return n * (n + 1) // 2

        n += 1


if __name__ == "__main__":
    print(solve(500))  # 76576500
